File: 0854-making-a-large-island/0854-making-a-large-island.java
class Solution {
    int[][] dir = {{0,1}, {0,-1}, {1,0}, {-1,0}};
    
    public int largestIsland(int[][] grid) {
        int R = grid.length, C = grid[0].length;
        int islandId = 2;
        Map<Integer, Integer> areaMap = new HashMap<>();

        for (int i = 0; i < R; i++) {
            for (int j = 0; j < C; j++) {
                if (grid[i][j] == 1) {
                    int area = dfs(grid, i, j, islandId);
                    areaMap.put(islandId, area);
                    islandId++;
                }
            }
        }

        int maxArea = areaMap.values().stream().max(Integer::compare).orElse(0);

        for (int i = 0; i < R; i++) {
            for (int j = 0; j < C; j++) {
                if (grid[i][j] == 0) {
                    Set<Integer> neighborIslands = new HashSet<>();
                    for (int[] d : dir) {
                        int ni = i + d[0], nj = j + d[1];
                        if (ni >= 0 && nj >= 0 && ni < R && nj < C && grid[ni][nj] > 1) {
                            neighborIslands.add(grid[ni][nj]);
                        }
                    }
                    int newArea = 1; 
                    for (int id : neighborIslands) {
                        newArea += areaMap.get(id);
                    }
                    maxArea = Math.max(maxArea, newArea);
                }
            }
        }
        
        return maxArea == 0 ? R * C : maxArea;
    }
    
    private int dfs(int[][] grid, int i, int j, int id) {
        if (i < 0 || j < 0 || i >= grid.length || j >= grid[0].length || grid[i][j] != 1) return 0;
        grid[i][j] = id;
        int area = 1;
        for (int[] d : dir) {
            area += dfs(grid, i + d[0], j + d[1], id);
        }
        return area;
    }
}

File: 0854-making-a-large-island/0854-making-a-large-island.cpp
class Solution {
public:
    int largestIsland(vector<vector<int>>& grid) {
        int n = grid.size();
        vector<vector<int>> directions = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
        unordered_map<int, int> area;
        int index = 2, max_area = 0;

        function<int(int, int, int)> dfs = [&](int r, int c, int id) {
            if (r < 0 || c < 0 || r >= n || c >= n || grid[r][c] != 1) return 0;
            grid[r][c] = id;
            int res = 1;
            for (auto& d : directions) {
                res += dfs(r + d[0], c + d[1], id);
            }
            return res;
        };

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (grid[i][j] == 1) {
                    area[index] = dfs(i, j, index);
                    max_area = max(max_area, area[index++]);
                }
            }
        }

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (grid[i][j] == 0) {
                    unordered_set<int> seen;
                    int new_area = 1;
                    for (auto& d : directions) {
                        int ni = i + d[0], nj = j + d[1];
                        if (ni >= 0 && nj >= 0 && ni < n && nj < n && grid[ni][nj] > 1) {
                            seen.insert(grid[ni][nj]);
                        }
                    }
                    for (int id : seen) new_area += area[id];
                    max_area = max(max_area, new_area);
                }
            }
        }

        return max_area;
    }
};
